gateway/Makefile

```diff
@@ -1,20 +1,22 @@
 include ../config.mk
 
-.PHONY: build push deploy
+.PHONY: envoy-xds-config envoy-config deploy-no-xds deploy
 
-GATEWAY_IMAGE := $(DOCKER_PREFIX)/gateway:$(TOKEN)
 IP := $(shell kubectl get secret global-config --template={{.data.ip}} | base64 --decode)
+SHA := $(shell git rev-parse --short=12 HEAD)
 
-export NAMESPACE := default
+envoy-xds-config:
+	HAIL_DOMAIN=$(DOMAIN) python3 ../ci/ci/envoy.py gateway ${HAIL}/letsencrypt/subdomains.txt ${HAIL}/gateway/cds.yaml ${HAIL}/gateway/rds.yaml
+	kubectl -n default create configmap gateway-xds-config \
+		--from-file=cds.yaml \
+		--from-file=rds.yaml \
+		--dry-run=client -o yaml \
+		| kubectl apply -f -
 
-build:
-	$(MAKE) -C ../docker hail-ubuntu
-	python3 ../ci/jinja2_render.py '{"hail_ubuntu_image":{"image":"'$$(cat ../docker/hail-ubuntu-image-ref)'"}}' Dockerfile Dockerfile.out
-	python3 ../ci/jinja2_render.py '{"domain": "$(DOMAIN)","subdomains":[$(shell sed -e 's/\(.*\)/"\1"/' ../letsencrypt/subdomains.txt | paste -s -d, -)]}' gateway.nginx.conf gateway.nginx.conf.out
-	../docker-build.sh . Dockerfile.out $(GATEWAY_IMAGE)
-
-deploy: build
-	python3 ../ci/jinja2_render.py '{"code":{"sha":"$(shell git rev-parse --short=12 HEAD)"},"global":{"ip":"$(IP)"}}' service.yaml service.yaml.out
+deploy:
```

Comment:
Why doesn't this depend on `envoy-xds-config`?

Reply:
Ya so this is a great question that maybe deserves better documentation (your comments below I think all get at this point). The `envoy-xds-config` target creates the yaml for defining the routes and clusters for production services. The reason I didn't make it a dependency of deploy is that if you're running CI, then running this target will temporary erase the routes for test namespaces (this target and CI both overwrite the same configmap). I don't think it's a big deal if that happens, CI rewrites the config every 10 seconds anyway so the worst thing that would happen is tests get stalled for a few seconds. But, I thought it best to be explicit about running that target anyway.

Reply:
For anyone *not* running test namespaces, running `envoy-xds-config` doesn't do any damage. The only time you'd need to re-run it is if you decided to add another service.

Reply:
Can we make this more idiot-proof with a warning or check of some sort? I'm just worried that all of our other services are `make deploy` and we're going to forget there's an extra target that might need to have been run. I guess if the config maps don't exist initially then the deploy will fail anyways. If that's the case, then this is fine. Otherwise, maybe we add an extra make target which combines both of those and name it with the intention for its use case. i.e. `deploy_with_config`?

Reply:
In other words, redeploying gateway and internal gateway already makes me nervous. I want to make sure I don't mess something up because the only time I'd probably need to do so is when the system is not under control.

Reply:
>  I guess if the config maps don't exist initially then the deploy will fail anyways. If that's the case, then this is fine.

Yup! This should fail because the deployment itself depends on these configmaps existing. So the way it would work in practice is:

- `make -C gateway deploy` without ever doing `make -C gateway envoy-xds-config` will fail because K8s will complain that `gateway-envoy-xds-config` doesn't exist.
- `make -C gateway deploy` won't ever *lose* any configuration.
- If you add a production service *and* you're not running CI, you need to do the extra step of running the `envoy-xds-config` target when you `make deploy`. This also has no danger of overwriting any test config because you're not running CI.


Also happy to add an additional target `deploy_with_config`, but I don't really see much of a difference between that and `make -C gateway envoy-xds-config deploy`.

Reply:
ok. I leave the decision in your hands. I think this is idiot proof enough for me.

Reply:
Ok, I've tried this a couple ways and I think this is the most reasonable option I could come up with for now. If there are complaints on ergonomics can definitely put in some time to make it smoother.